### ISSUE TITLE
ESLint cache not invalidated after removing --report-unused-disable-directives

### ISSUE THREAD

**Issue by edmorley** [eslint/eslint#13505]
**Tell us about your environment**

* **ESLint Version:** v7.5.0
* **Node Version:** v14.5.0
* **~npm~ Yarn Version:** v1.22.4

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
// .eslintrc.js
module.exports = {
  "rules": {
    "semi": "error",
  },
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
// eslint-disable-line semi
var a = 1;

```

```bash
yarn eslint --cache --report-unused-disable-directives .
// followed by
yarn eslint --cache .
```

**What did you expect to happen?**
On the second ESLint invocation, no error message should be printed about the unused disable directive, since that flag was not passed in on the CLI, and so the cached result from the first run should not be re-used.

**What actually happened? Please include the actual, raw output from ESLint.**
On the second ESLint invocation the unused disable directive error was still shown.
Performing an `rm .eslintcache` was required to make the error go away.

```
testcase $ yarn eslint --cache --report-unused-disable-directives .
yarn run v1.22.4
$ /Users/emorley/src/testcase/node_modules/.bin/eslint --cache --report-unused-disable-directives .

/Users/emorley/src/testcase/test.js
  1:1  error  Unused eslint-disable directive (no problems were reported from 'semi')

✖ 1 problem (1 error, 0 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

For the second invocation...
```
testcase $ yarn eslint --cache .
yarn run v1.22.4
$ /Users/emorley/src/testcase/node_modules/.bin/eslint --cache .

/Users/emorley/src/testcase/test.js
  1:1  error  Unused eslint-disable directive (no problems were reported from 'semi')

✖ 1 problem (1 error, 0 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Are you willing to submit a pull request to fix this bug?**
If someone could point to a few places in the source where this is handled, possibly :-)

**Comment by mdjermanovic** [comment]
Hi @edmorley, thanks for the bug report!

I can confirm that this is happening and that the final result is incorrect.

We might account for those cli options that will be passed to the linter, either in the [config hash](https://github.com/eslint/eslint/blob/f3a19d2a7ea9505000d14229a450dba133c10d5e/lib/cli-engine/lint-result-cache.js#L32) or maybe store an additional hash.

Only, that fix might have some performance costs for all users and I'm not sure how common this particular use case is.
